### PR TITLE
Add size assert, refactor

### PR DESCRIPTION
**Pull request by gordicaleksa**
Minor refactor + added size assert.

`inline` was causing compile failures on my Linux machine, i think if we want to make it inline we'd have to move it to the header file (but then again see my next comment on simplicity vs robustness tradeoff so if we err on the simplicity side i say let's just omit this minor optimization (?)).

### CODE REVIEW

**Comment by gordicaleksa** [comment]
@karpathy similarly to the circular buffer logic PR from ngram repo (see https://github.com/EurekaLabsAI/ngram/pull/10) - happy to modify `tensor_to_string` to make it more flexible/dynamic akin to how Python's list or C++ std::vector work (when we hit the limit of the underlaying buffer reallocate a 2x larger buffer and copy the elements over).

Having said that it's a tradeoff between simplicity and robustness. Lmk if you'd like to see that change.

File: tensor1d.c [hunk 1]
```diff
@@ -157,11 +158,9 @@ Tensor* tensor_slice(Tensor* t, int start, int end, int step) {
     // 1) handle negative indices by wrapping around
     if (start < 0) { start = t->size + start; }
     if (end < 0) { end = t->size + end; }
-    // 2) handle out-of-bounds indices: clip to 0 and t->size
-    if (start < 0) { start = 0; }
-    if (end < 0) { end = 0; }
-    if (start > t->size) { start = t->size; }
-    if (end > t->size) { end = t->size; }
+    // 2) handle out-of-bounds indices: clip to [0, t->size] range
+    start = fmin(fmax(start, 0), t->size);
```
Comment (karpathy) [inline tensor1d.c:162]: The `fmin` and `fmax` in C are on `double`, so there will be a weird cast from int to double here, would prefer to not do that. An alternative to the ifs would possibly be a ternary expression, I'd be ok with that

https://www.educative.io/answers/what-is-fmin-in-c

File: tensor1d.c [hunk 2]
```diff
@@ -41,6 +41,7 @@ inline int ceil_div(int a, int b) {
 // similar to torch.Storage
 
 Storage* storage_new(int size) {
+    assert(size > 0);
```
Comment (karpathy) [inline tensor1d.c:44]: good call

File: tensor1d.c [hunk 3]
```diff
@@ -30,7 +30,7 @@ void *malloc_check(size_t size, const char *file, int line) {
 // ----------------------------------------------------------------------------
 // utils
 
-inline int ceil_div(int a, int b) {
```
Comment (karpathy) [inline tensor1d.c:33]: agree on getting rid of inline if it causes issues. I didn't realize `inline` was contentious and has sharp edges.

File: tensor1d.c (hunk 1 above)
Comment (gordicaleksa) [reply tensor1d.c]: @karpathy ok good point, my bad here -> fixed, see whether you like this solution

**Pull request merged**